Gemini - Consistent Error "Expected at least one candidate in AI Studio API response."

**Describe the bug**
Using `google:gemini-2.5-flash` model I keep getting this error. If I switch to 1.5 (`google:gemini-1.5-flash`) the errors don't seem to occur.

```
[ERROR] Error: Expected at least one candidate in AI Studio API response.
```

**To Reproduce**

Its been occurring quite consistently. This is probably more complicated example than needed - but 2 of the 3 requests failed for me (which is especially strange)

```
[ERROR] Error: Expected at least one candidate in AI Studio API response.
```

```prompt.json
[
  {
    "role": "system",
    "content": "{{ system_message }}"
  }
  {% for message in conversation %}
  ,
  {
    "role": "{{ message.role }}",
    "content": "{{ message.content }}"
  }
  {% endfor %}
]
```

```yaml
# promptfooconfig.yaml

prompts:
  - file://prompt.json

providers:
  - google:gemini-2.5-flash

defaultTest:
  vars:
    system_message: "This is your single large system prompt. You are a helpful assistant that provides concise and accurate information. You should be friendly and professional."

tests:
  - description: "Conversation about geography"
    vars:
      conversation:
        - role: user
          content: "What is the capital of France?"
        - role: assistant
          content: "The capital of France is Paris."
        - role: user
          content: "What is its population?"

  - description: "Conversation about literature"
    vars:
      conversation:
        - role: user
          content: "Who wrote 'To Kill a Mockingbird'?"
        - role: assistant
          content: "Harper Lee wrote 'To Kill a Mockingbird'."
        - role: user
          content: "What is the main theme of the book?"

  - description: "A shorter conversation"
    vars:
      conversation:
        - role: user
          content: "What is the formula for water?"
```

**Expected behavior**

The prompts are simple and (should be) straight forward - there should be no error occuring.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**System information:**

```
===================================================================================================================================
Promptfoo Debug Information
===================================================================================================================================
{
  "version": "0.117.10",
  "platform": {
    "os": "darwin",
    "release": "24.3.0",
    "arch": "arm64",
    "nodeVersion": "v22.14.0"
  },
  "env": {},
  "configInfo": {
    "defaultConfigPath": "/Users/gb/github-gigpro/poc-voiceflow/promptfooconfig.yaml",
    "configExists": true,
    "configContent": {
      "config": {
        "tags": {},
        "prompts": [
          "file:///Users/gb/github-gigpro/poc-voiceflow/prompt.json"
        ],
        "providers": [
          "google:gemini-1.5-flash"
        ],
        "tests": [
          {
            "description": "Conversation about geography",
            "vars": {
              "conversation": [
                {
                  "role": "user",
                  "content": "What is the capital of France?"
                },
                {
                  "role": "assistant",
                  "content": "The capital of France is Paris."
                },
                {
                  "role": "user",
                  "content": "What is its population?"
                }
              ]
            }
          },
          {
            "description": "Conversation about literature",
            "vars": {
              "conversation": [
                {
                  "role": "user",
                  "content": "Who wrote 'To Kill a Mockingbird'?"
                },
                {
                  "role": "assistant",
                  "content": "Harper Lee wrote 'To Kill a Mockingbird'."
                },
                {
                  "role": "user",
                  "content": "What is the main theme of the book?"
                }
              ]
            }
          },
          {
            "description": "A shorter conversation",
            "vars": {
              "conversation": [
                {
                  "role": "user",
                  "content": "What is the formula for water?"
                }
              ]
            }
          }
        ],
        "scenarios": [],
        "env": {},
        "sharing": true,
        "defaultTest": {
          "vars": {
            "system_message": "file://system_prompt.txt"
          },
          "assert": [],
          "options": {},
          "metadata": {}
        },
        "outputPath": [],
        "extensions": [],
        "metadata": {}
      },
      "testSuite": {
        "tags": {},
        "prompts": [
          {
            "raw": "[\n  {\n    \"role\": \"system\",\n    \"content\": \"{{ system_message }}\"\n  }\n  {% for message in conversation %}\n  ,\n  {\n    \"role\": \"{{ message.role }}\",\n    \"content\": \"{{ message.content }}\"\n  }\n  {% endfor %}\n]",
            "label": "/Users/gb/github-gigpro/poc-voiceflow/prompt.json: [\n  {\n    \"role\": \"system\",\n    \"content\": \"{{ system_message }}\"\n  }\n  {% for message in conversation %}\n  ,\n  {\n    \"role\": \"{{ message.role }}\",\n    \"content\": \"{{ message.content }}\"\n  }\n  {% endfor %}\n]"
          }
        ],
        "providers": [
          {
            "env": {},
            "modelName": "gemini-1.5-flash",
            "config": {
              "basePath": "/Users/gb/github-gigpro/poc-voiceflow"
            },
            "mcpClient": null,
            "initializationPromise": null,
            "label": ""
          }
        ],
        "providerPromptMap": {},
        "tests": [
          {
            "description": "Conversation about geography",
            "vars": {
              "conversation": [
                {
                  "role": "user",
                  "content": "What is the capital of France?"
                },
                {
                  "role": "assistant",
                  "content": "The capital of France is Paris."
                },
                {
                  "role": "user",
                  "content": "What is its population?"
                }
              ]
            }
          },
          {
            "description": "Conversation about literature",
            "vars": {
              "conversation": [
                {
                  "role": "user",
                  "content": "Who wrote 'To Kill a Mockingbird'?"
                },
                {
                  "role": "assistant",
                  "content": "Harper Lee wrote 'To Kill a Mockingbird'."
                },
                {
                  "role": "user",
                  "content": "What is the main theme of the book?"
                }
              ]
            }
          },
          {
            "description": "A shorter conversation",
            "vars": {
              "conversation": [
                {
                  "role": "user",
                  "content": "What is the formula for water?"
                }
              ]
            }
          }
        ],
        "scenarios": [],
        "defaultTest": {
          "metadata": {},
          "options": {},
          "vars": {
            "system_message": "file://system_prompt.txt"
          },
          "assert": []
        },
        "nunjucksFilters": {},
        "extensions": []
      },
      "basePath": "/Users/gb/github-gigpro/poc-voiceflow"
    }
  }
}
```

- Promptfoo version:  0.117.10

**Additional context**

The error has been occurring quite consistently with Gemini. I'd imagine its trying to make some unexpected call or response format.

--- a/site/docs/providers/google.md
+++ b/site/docs/providers/google.md
@@ -457,6 +457,24 @@ providers:
 
 System instructions support Nunjucks templating and can be loaded from external files for better organization and reusability.
 
+### Role Mapping Configuration
+
+Gemini models require specific role names in chat messages. By default, Promptfoo uses the `model` role for compatibility with newer Gemini versions (2.5+). For older Gemini versions that expect the `assistant` role, you can disable this:
+
+```yaml
+providers:
+  # Default behavior - maps 'assistant' to 'model' (for Gemini 2.5+)
+  - id: google:gemini-2.5-flash
+    config:
+      temperature: 0.7
+
+  # For older Gemini versions - preserve 'assistant' role
+  - id: google:gemini-1.5-pro
+    config:
+      useAssistantRole: true # Preserves 'assistant' role without mapping
+      temperature: 0.7
+```
+
 For more details on capabilities and configuration options, see the [Gemini API documentation](https://ai.google.dev/docs).
 
 ## Model Examples

--- a/src/providers/google/ai.studio.ts
+++ b/src/providers/google/ai.studio.ts
@@ -258,6 +258,7 @@ export class AIStudioChatProvider extends AIStudioGenericProvider {
       prompt,
       context?.vars,
       config.systemInstruction,
+      { useAssistantRole: config.useAssistantRole },
     );
 
     // Determine API version based on model

--- a/src/providers/google/live.ts
+++ b/src/providers/google/live.ts
@@ -108,6 +108,7 @@ export class GoogleLiveProvider implements ApiProvider {
       prompt,
       context?.vars,
       this.config.systemInstruction,
+      { useAssistantRole: this.config.useAssistantRole },
     );
     let contentIndex = 0;
 

--- a/src/providers/google/types.ts
+++ b/src/providers/google/types.ts
@@ -219,6 +219,13 @@ export interface CompletionOptions {
   };
 
   mcp?: MCPConfig;
+
+  /**
+   * Controls role mapping when converting from OpenAI format to Gemini format.
+   * If true, uses 'assistant' role (for older Gemini versions).
+   * If false (default), maps 'assistant' to 'model' (for newer Gemini versions).
+   */
+  useAssistantRole?: boolean;
 }
 
 // Claude API interfaces

--- a/src/providers/google/util.ts
+++ b/src/providers/google/util.ts
@@ -197,37 +197,20 @@ export function maybeCoerceToGeminiFormat(
     contents.every((item) => typeof item.content === 'string')
   ) {
     // This looks like an OpenAI chat format
-    // Map 'assistant' to 'model' for Gemini API compatibility
-    // The useAssistantRole flag is deprecated - we always map to 'model' now
-    coercedContents = contents.map((item) => {
-      let mappedRole: 'user' | 'model' | 'system' | undefined;
-      if (item.role === 'assistant') {
-        // Always map assistant to model (Gemini API only accepts 'user' or 'model')
-        mappedRole = 'model';
-      } else if (item.role === 'user' || item.role === 'model' || item.role === 'system') {
-        mappedRole = item.role as 'user' | 'model' | 'system';
-      } else {
-        mappedRole = undefined;
-      }
-      return {
-        role: mappedRole as 'user' | 'model' | undefined,
-        parts: [{ text: item.content }],
-      };
-    });
+    const targetRole = options?.useAssistantRole ? 'assistant' : 'model';
+    coercedContents = contents.map((item) => ({
+      role: (item.role === 'assistant' ? targetRole : item.role) as 'user' | 'model' | undefined,
+      parts: [{ text: item.content }],
+    }));
     coerced = true;
   } else if (Array.isArray(contents) && contents.every((item) => item.role && item.content)) {
     // This looks like an OpenAI chat format with content that might be an array or object
+    const targetRole = options?.useAssistantRole ? 'assistant' : 'model';
     coercedContents = contents.map((item) => {
-      // Always map 'assistant' to 'model' for Gemini API compatibility
-      let mappedRole: 'user' | 'model' | 'system' | undefined;
-      if (item.role === 'assistant') {
-        // Always map assistant to model (Gemini API only accepts 'user' or 'model')
-        mappedRole = 'model';
-      } else if (item.role === 'user' || item.role === 'model' || item.role === 'system') {
-        mappedRole = item.role as 'user' | 'model' | 'system';
-      } else {
-        mappedRole = undefined;
-      }
+      const mappedRole = (item.role === 'assistant' ? targetRole : item.role) as
+        | 'user'
+        | 'model'
+        | undefined;
       if (Array.isArray(item.content)) {
         // Handle array content
         const parts = item.content.map((contentItem: any) => {
@@ -241,19 +224,19 @@ export function maybeCoerceToGeminiFormat(
           }
         });
         return {
-          role: mappedRole as 'user' | 'model' | undefined,
+          role: mappedRole,
           parts,
         };
       } else if (typeof item.content === 'object') {
         // Handle object content
         return {
-          role: mappedRole as 'user' | 'model' | undefined,
+          role: mappedRole,
           parts: [item.content],
         };
       } else {
         // Handle string content
         return {
-          role: mappedRole as 'user' | 'model' | undefined,
+          role: mappedRole,
           parts: [{ text: item.content }],
         };
       }

--- a/src/providers/google/util.ts
+++ b/src/providers/google/util.ts
@@ -130,7 +130,10 @@ const GeminiFormatSchema = z.array(ContentSchema);
 
 export type GeminiFormat = z.infer<typeof GeminiFormatSchema>;
 
-export function maybeCoerceToGeminiFormat(contents: any): {
+export function maybeCoerceToGeminiFormat(
+  contents: any,
+  options?: { useAssistantRole?: boolean },
+): {
   contents: GeminiFormat;
   coerced: boolean;
   systemInstruction: { parts: [Part, ...Part[]] } | undefined;
@@ -194,14 +197,20 @@ export function maybeCoerceToGeminiFormat(contents: any): {
     contents.every((item) => typeof item.content === 'string')
   ) {
     // This looks like an OpenAI chat format
+    const targetRole = options?.useAssistantRole ? 'assistant' : 'model';
     coercedContents = contents.map((item) => ({
-      role: item.role as 'user' | 'model' | undefined,
+      role: (item.role === 'assistant' ? targetRole : item.role) as 'user' | 'model' | undefined,
       parts: [{ text: item.content }],
     }));
     coerced = true;
   } else if (Array.isArray(contents) && contents.every((item) => item.role && item.content)) {
     // This looks like an OpenAI chat format with content that might be an array or object
+    const targetRole = options?.useAssistantRole ? 'assistant' : 'model';
     coercedContents = contents.map((item) => {
+      const mappedRole = (item.role === 'assistant' ? targetRole : item.role) as
+        | 'user'
+        | 'model'
+        | undefined;
       if (Array.isArray(item.content)) {
         // Handle array content
         const parts = item.content.map((contentItem: any) => {
@@ -215,19 +224,19 @@ export function maybeCoerceToGeminiFormat(contents: any): {
           }
         });
         return {
-          role: item.role as 'user' | 'model' | undefined,
+          role: mappedRole,
           parts,
         };
       } else if (typeof item.content === 'object') {
         // Handle object content
         return {
-          role: item.role as 'user' | 'model' | undefined,
+          role: mappedRole,
           parts: [item.content],
         };
       } else {
         // Handle string content
         return {
-          role: item.role as 'user' | 'model' | undefined,
+          role: mappedRole,
           parts: [{ text: item.content }],
         };
       }
@@ -355,11 +364,35 @@ export async function hasGoogleDefaultCredentials() {
 
 export function getCandidate(data: GeminiResponseData) {
   if (!data || !data.candidates || data.candidates.length < 1) {
-    throw new Error('Expected at least one candidate in AI Studio API response.');
+    // Check if the prompt was blocked
+    let errorDetails = 'No candidates returned in API response.';
+
+    if (data?.promptFeedback?.blockReason) {
+      errorDetails = `Response blocked: ${data.promptFeedback.blockReason}`;
+      if (data.promptFeedback.safetyRatings) {
+        const flaggedCategories = data.promptFeedback.safetyRatings
+          .filter((rating) => rating.probability !== 'NEGLIGIBLE')
+          .map((rating) => `${rating.category}: ${rating.probability}`);
+        if (flaggedCategories.length > 0) {
+          errorDetails += ` (Safety ratings: ${flaggedCategories.join(', ')})`;
+        }
+      }
+    } else if (data?.promptFeedback?.safetyRatings) {
+      const flaggedCategories = data.promptFeedback.safetyRatings
+        .filter((rating) => rating.probability !== 'NEGLIGIBLE')
+        .map((rating) => `${rating.category}: ${rating.probability}`);
+      if (flaggedCategories.length > 0) {
+        errorDetails = `Response may have been blocked due to safety filters: ${flaggedCategories.join(', ')}`;
+      }
+    }
+
+    errorDetails += `\n\nGot response: ${JSON.stringify(data)}`;
+
+    throw new Error(errorDetails);
   }
   if (data.candidates.length > 1) {
     logger.debug(
-      `Expected one candidate in AI Studio API response, but got ${data.candidates.length}.`,
+      `Expected one candidate in AI Studio API response, but got ${data.candidates.length}: ${JSON.stringify(data)}`,
     );
   }
   const candidate = data.candidates[0];
@@ -367,6 +400,38 @@ export function getCandidate(data: GeminiResponseData) {
 }
 
 export function formatCandidateContents(candidate: Candidate) {
+  // Check if the candidate was blocked or stopped for safety reasons
+  if (
+    candidate.finishReason &&
+    ['SAFETY', 'RECITATION', 'PROHIBITED_CONTENT', 'BLOCKLIST', 'SPII'].includes(
+      candidate.finishReason,
+    )
+  ) {
+    let errorMessage = `Response was blocked with finish reason: ${candidate.finishReason}`;
+
+    if (candidate.safetyRatings) {
+      const flaggedCategories = candidate.safetyRatings
+        .filter((rating) => rating.probability !== 'NEGLIGIBLE' || rating.blocked)
+        .map(
+          (rating) =>
+            `${rating.category}: ${rating.probability}${rating.blocked ? ' (BLOCKED)' : ''}`,
+        );
+      if (flaggedCategories.length > 0) {
+        errorMessage += `\nSafety ratings: ${flaggedCategories.join(', ')}`;
+      }
+    }
+
+    if (candidate.finishReason === 'RECITATION') {
+      errorMessage +=
+        "\n\nThis typically occurs when the response is too similar to content from the model's training data.";
+    } else if (candidate.finishReason === 'SAFETY') {
+      errorMessage +=
+        '\n\nThe response was blocked due to safety filters. Consider adjusting safety settings or modifying your prompt.';
+    }
+
+    throw new Error(errorMessage);
+  }
+
   if (candidate.content?.parts) {
     let output = '';
     let is_text = true;
@@ -590,6 +655,7 @@ export function geminiFormatAndSystemInstructions(
   prompt: string,
   contextVars?: Record<string, string | object>,
   configSystemInstruction?: Content | string,
+  options?: { useAssistantRole?: boolean },
 ): {
   contents: GeminiFormat;
   systemInstruction: Content | { parts: [Part, ...Part[]] } | undefined;
@@ -608,7 +674,7 @@ export function geminiFormatAndSystemInstructions(
     contents: updatedContents,
     coerced,
     systemInstruction: parsedSystemInstruction,
-  } = maybeCoerceToGeminiFormat(contents);
+  } = maybeCoerceToGeminiFormat(contents, options);
   if (coerced) {
     logger.debug(`Coerced JSON prompt to Gemini format: ${JSON.stringify(contents)}`);
     contents = updatedContents;

--- a/src/providers/google/util.ts
+++ b/src/providers/google/util.ts
@@ -197,20 +197,37 @@ export function maybeCoerceToGeminiFormat(
     contents.every((item) => typeof item.content === 'string')
   ) {
     // This looks like an OpenAI chat format
-    const targetRole = options?.useAssistantRole ? 'assistant' : 'model';
-    coercedContents = contents.map((item) => ({
-      role: (item.role === 'assistant' ? targetRole : item.role) as 'user' | 'model' | undefined,
-      parts: [{ text: item.content }],
-    }));
+    // Map 'assistant' to 'model' for Gemini API compatibility
+    // The useAssistantRole flag is deprecated - we always map to 'model' now
+    coercedContents = contents.map((item) => {
+      let mappedRole: 'user' | 'model' | 'system' | undefined;
+      if (item.role === 'assistant') {
+        // Always map assistant to model (Gemini API only accepts 'user' or 'model')
+        mappedRole = 'model';
+      } else if (item.role === 'user' || item.role === 'model' || item.role === 'system') {
+        mappedRole = item.role as 'user' | 'model' | 'system';
+      } else {
+        mappedRole = undefined;
+      }
+      return {
+        role: mappedRole as 'user' | 'model' | undefined,
+        parts: [{ text: item.content }],
+      };
+    });
     coerced = true;
   } else if (Array.isArray(contents) && contents.every((item) => item.role && item.content)) {
     // This looks like an OpenAI chat format with content that might be an array or object
-    const targetRole = options?.useAssistantRole ? 'assistant' : 'model';
     coercedContents = contents.map((item) => {
-      const mappedRole = (item.role === 'assistant' ? targetRole : item.role) as
-        | 'user'
-        | 'model'
-        | undefined;
+      // Always map 'assistant' to 'model' for Gemini API compatibility
+      let mappedRole: 'user' | 'model' | 'system' | undefined;
+      if (item.role === 'assistant') {
+        // Always map assistant to model (Gemini API only accepts 'user' or 'model')
+        mappedRole = 'model';
+      } else if (item.role === 'user' || item.role === 'model' || item.role === 'system') {
+        mappedRole = item.role as 'user' | 'model' | 'system';
+      } else {
+        mappedRole = undefined;
+      }
       if (Array.isArray(item.content)) {
         // Handle array content
         const parts = item.content.map((contentItem: any) => {
@@ -224,19 +241,19 @@ export function maybeCoerceToGeminiFormat(
           }
         });
         return {
-          role: mappedRole,
+          role: mappedRole as 'user' | 'model' | undefined,
           parts,
         };
       } else if (typeof item.content === 'object') {
         // Handle object content
         return {
-          role: mappedRole,
+          role: mappedRole as 'user' | 'model' | undefined,
           parts: [item.content],
         };
       } else {
         // Handle string content
         return {
-          role: mappedRole,
+          role: mappedRole as 'user' | 'model' | undefined,
           parts: [{ text: item.content }],
         };
       }

--- a/src/providers/google/vertex.ts
+++ b/src/providers/google/vertex.ts
@@ -296,6 +296,7 @@ export class VertexChatProvider extends VertexGenericProvider {
       prompt,
       context?.vars,
       config.systemInstruction,
+      { useAssistantRole: config.useAssistantRole },
     );
     // --- MCP tool injection logic ---
     const mcpTools = this.mcpClient ? transformMCPToolsToGoogle(this.mcpClient.getAllTools()) : [];

--- a/test/providers/google/ai.studio.test.ts
+++ b/test/providers/google/ai.studio.test.ts
@@ -217,8 +217,7 @@ describe('AIStudioChatProvider', () => {
 
       const response = await provider.callGemini('test prompt');
       expect(response.error).toContain('No candidates returned in API response');
-      expect(response.error).toContain('Common causes:');
-      expect(response.error).toContain('Safety filters blocking content');
+      expect(response.error).toContain('Got response:');
     });
 
     it('should handle malformed API responses', async () => {
@@ -292,7 +291,9 @@ describe('AIStudioChatProvider', () => {
       const response = await provider.callGemini('test prompt');
       expect(response.error).toContain('Response blocked: PROHIBITED_CONTENT');
       expect(response.error).toContain('HARM_CATEGORY_HATE_SPEECH: HIGH');
-      expect(response.error).not.toContain('HARM_CATEGORY_HARASSMENT'); // NEGLIGIBLE should be filtered out
+      // NEGLIGIBLE should be filtered out from the safety ratings summary
+      expect(response.error).toMatch(/Safety ratings: HARM_CATEGORY_HATE_SPEECH: HIGH\)/);
+      expect(response.error).not.toMatch(/Safety ratings:.*HARM_CATEGORY_HARASSMENT.*HIGH/); // Only check it's not in the summary with HIGH
     });
 
     it('should handle candidates blocked with finish reason', async () => {

--- a/test/providers/google/ai.studio.test.ts
+++ b/test/providers/google/ai.studio.test.ts
@@ -216,9 +216,9 @@ describe('AIStudioChatProvider', () => {
       });
 
       const response = await provider.callGemini('test prompt');
-      expect(response.error).toContain(
-        'Error: Expected at least one candidate in AI Studio API response.',
-      );
+      expect(response.error).toContain('No candidates returned in API response');
+      expect(response.error).toContain('Common causes:');
+      expect(response.error).toContain('Safety filters blocking content');
     });
 
     it('should handle malformed API responses', async () => {
@@ -257,6 +257,83 @@ describe('AIStudioChatProvider', () => {
       expect(response).toEqual({
         error: 'Error: No output found in response: {"content":{"parts":null}}',
       });
+    });
+
+    it('should handle responses blocked with promptFeedback', async () => {
+      const provider = new AIStudioChatProvider('gemini-pro', {
+        config: {
+          apiKey: 'test-key',
+        },
+      });
+
+      jest.mocked(util.maybeCoerceToGeminiFormat).mockReturnValueOnce({
+        contents: [{ role: 'user', parts: [{ text: 'test prompt' }] }],
+        coerced: false,
+        systemInstruction: undefined,
+      });
+
+      jest.mocked(cache.fetchWithCache).mockResolvedValueOnce({
+        data: {
+          candidates: [],
+          promptFeedback: {
+            blockReason: 'PROHIBITED_CONTENT',
+            safetyRatings: [
+              { category: 'HARM_CATEGORY_HATE_SPEECH', probability: 'HIGH' },
+              { category: 'HARM_CATEGORY_HARASSMENT', probability: 'NEGLIGIBLE' },
+            ],
+          },
+        },
+        cached: false,
+        status: 200,
+        statusText: 'OK',
+        headers: {},
+      });
+
+      const response = await provider.callGemini('test prompt');
+      expect(response.error).toContain('Response blocked: PROHIBITED_CONTENT');
+      expect(response.error).toContain('HARM_CATEGORY_HATE_SPEECH: HIGH');
+      expect(response.error).not.toContain('HARM_CATEGORY_HARASSMENT'); // NEGLIGIBLE should be filtered out
+    });
+
+    it('should handle candidates blocked with finish reason', async () => {
+      const provider = new AIStudioChatProvider('gemini-pro', {
+        config: {
+          apiKey: 'test-key',
+        },
+      });
+
+      jest.mocked(util.maybeCoerceToGeminiFormat).mockReturnValueOnce({
+        contents: [{ role: 'user', parts: [{ text: 'test prompt' }] }],
+        coerced: false,
+        systemInstruction: undefined,
+      });
+
+      jest.mocked(cache.fetchWithCache).mockResolvedValueOnce({
+        data: {
+          candidates: [
+            {
+              content: { parts: [{ text: '' }] },
+              finishReason: 'RECITATION',
+              safetyRatings: [
+                {
+                  category: 'HARM_CATEGORY_DANGEROUS_CONTENT',
+                  probability: 'MEDIUM',
+                  blocked: false,
+                },
+              ],
+            },
+          ],
+        },
+        cached: false,
+        status: 200,
+        statusText: 'OK',
+        headers: {},
+      });
+
+      const response = await provider.callGemini('test prompt');
+      expect(response.error).toContain('Response was blocked with finish reason: RECITATION');
+      expect(response.error).toContain("too similar to content from the model's training data");
+      expect(response.error).toContain('HARM_CATEGORY_DANGEROUS_CONTENT: MEDIUM');
     });
   });
 

--- a/test/providers/google/util.test.ts
+++ b/test/providers/google/util.test.ts
@@ -270,7 +270,7 @@ describe('util', () => {
       });
     });
 
-    it('should map assistant role to model even when useAssistantRole is true (Gemini API constraint)', () => {
+    it('should preserve assistant role when useAssistantRole is true', () => {
       const input = [
         { role: 'user', content: 'What is the capital of France?' },
         { role: 'assistant', content: 'The capital of France is Paris.' },
@@ -280,7 +280,7 @@ describe('util', () => {
       expect(result).toEqual({
         contents: [
           { role: 'user', parts: [{ text: 'What is the capital of France?' }] },
-          { role: 'model', parts: [{ text: 'The capital of France is Paris.' }] }, // Always maps to model
+          { role: 'assistant', parts: [{ text: 'The capital of France is Paris.' }] }, // assistant preserved
           { role: 'user', parts: [{ text: 'What is its population?' }] },
         ],
         coerced: true,
@@ -324,17 +324,17 @@ describe('util', () => {
       });
     });
 
-    it('should always map assistant to model regardless of useAssistantRole flag (Gemini API constraint)', () => {
+    it('should respect useAssistantRole flag with array content', () => {
       const input = [
         { role: 'user', content: [{ type: 'text', text: 'Question' }] },
         { role: 'assistant', content: [{ type: 'text', text: 'Answer' }] },
       ];
 
-      // Test with useAssistantRole: true - should still map to model
+      // Test with useAssistantRole: true
       const resultWithAssistant = maybeCoerceToGeminiFormat(input, { useAssistantRole: true });
-      expect(resultWithAssistant.contents[1].role).toBe('model');
+      expect(resultWithAssistant.contents[1].role).toBe('assistant');
 
-      // Test with useAssistantRole: false (default) - should also map to model
+      // Test with useAssistantRole: false (default)
       const resultWithModel = maybeCoerceToGeminiFormat(input, { useAssistantRole: false });
       expect(resultWithModel.contents[1].role).toBe('model');
     });

--- a/test/providers/google/util.test.ts
+++ b/test/providers/google/util.test.ts
@@ -270,7 +270,7 @@ describe('util', () => {
       });
     });
 
-    it('should preserve assistant role when useAssistantRole is true', () => {
+    it('should map assistant role to model even when useAssistantRole is true (Gemini API constraint)', () => {
       const input = [
         { role: 'user', content: 'What is the capital of France?' },
         { role: 'assistant', content: 'The capital of France is Paris.' },
@@ -280,7 +280,7 @@ describe('util', () => {
       expect(result).toEqual({
         contents: [
           { role: 'user', parts: [{ text: 'What is the capital of France?' }] },
-          { role: 'assistant', parts: [{ text: 'The capital of France is Paris.' }] }, // assistant preserved
+          { role: 'model', parts: [{ text: 'The capital of France is Paris.' }] }, // Always maps to model
           { role: 'user', parts: [{ text: 'What is its population?' }] },
         ],
         coerced: true,
@@ -324,17 +324,17 @@ describe('util', () => {
       });
     });
 
-    it('should respect useAssistantRole flag with array content', () => {
+    it('should always map assistant to model regardless of useAssistantRole flag (Gemini API constraint)', () => {
       const input = [
         { role: 'user', content: [{ type: 'text', text: 'Question' }] },
         { role: 'assistant', content: [{ type: 'text', text: 'Answer' }] },
       ];
 
-      // Test with useAssistantRole: true
+      // Test with useAssistantRole: true - should still map to model
       const resultWithAssistant = maybeCoerceToGeminiFormat(input, { useAssistantRole: true });
-      expect(resultWithAssistant.contents[1].role).toBe('assistant');
+      expect(resultWithAssistant.contents[1].role).toBe('model');
 
-      // Test with useAssistantRole: false (default)
+      // Test with useAssistantRole: false (default) - should also map to model
       const resultWithModel = maybeCoerceToGeminiFormat(input, { useAssistantRole: false });
       expect(resultWithModel.contents[1].role).toBe('model');
     });

--- a/test/providers/google/util.test.ts
+++ b/test/providers/google/util.test.ts
@@ -252,6 +252,58 @@ describe('util', () => {
       });
     });
 
+    it('should map assistant role to model role by default', () => {
+      const input = [
+        { role: 'user', content: 'What is the capital of France?' },
+        { role: 'assistant', content: 'The capital of France is Paris.' },
+        { role: 'user', content: 'What is its population?' },
+      ];
+      const result = maybeCoerceToGeminiFormat(input);
+      expect(result).toEqual({
+        contents: [
+          { role: 'user', parts: [{ text: 'What is the capital of France?' }] },
+          { role: 'model', parts: [{ text: 'The capital of France is Paris.' }] }, // assistant mapped to model
+          { role: 'user', parts: [{ text: 'What is its population?' }] },
+        ],
+        coerced: true,
+        systemInstruction: undefined,
+      });
+    });
+
+    it('should preserve assistant role when useAssistantRole is true', () => {
+      const input = [
+        { role: 'user', content: 'What is the capital of France?' },
+        { role: 'assistant', content: 'The capital of France is Paris.' },
+        { role: 'user', content: 'What is its population?' },
+      ];
+      const result = maybeCoerceToGeminiFormat(input, { useAssistantRole: true });
+      expect(result).toEqual({
+        contents: [
+          { role: 'user', parts: [{ text: 'What is the capital of France?' }] },
+          { role: 'assistant', parts: [{ text: 'The capital of France is Paris.' }] }, // assistant preserved
+          { role: 'user', parts: [{ text: 'What is its population?' }] },
+        ],
+        coerced: true,
+        systemInstruction: undefined,
+      });
+    });
+
+    it('should map assistant to model when useAssistantRole is false', () => {
+      const input = [
+        { role: 'user', content: 'Hello' },
+        { role: 'assistant', content: 'Hi there' },
+      ];
+      const result = maybeCoerceToGeminiFormat(input, { useAssistantRole: false });
+      expect(result).toEqual({
+        contents: [
+          { role: 'user', parts: [{ text: 'Hello' }] },
+          { role: 'model', parts: [{ text: 'Hi there' }] },
+        ],
+        coerced: true,
+        systemInstruction: undefined,
+      });
+    });
+
     it('should handle OpenAI chat format with array content', () => {
       const input = [
         {
@@ -270,6 +322,21 @@ describe('util', () => {
         coerced: true,
         systemInstruction: undefined,
       });
+    });
+
+    it('should respect useAssistantRole flag with array content', () => {
+      const input = [
+        { role: 'user', content: [{ type: 'text', text: 'Question' }] },
+        { role: 'assistant', content: [{ type: 'text', text: 'Answer' }] },
+      ];
+
+      // Test with useAssistantRole: true
+      const resultWithAssistant = maybeCoerceToGeminiFormat(input, { useAssistantRole: true });
+      expect(resultWithAssistant.contents[1].role).toBe('assistant');
+
+      // Test with useAssistantRole: false (default)
+      const resultWithModel = maybeCoerceToGeminiFormat(input, { useAssistantRole: false });
+      expect(resultWithModel.contents[1].role).toBe('model');
     });
 
     it('should handle OpenAI chat format with object content', () => {

